### ISSUE TITLE
Add support for specifying the PathBase for all requests

### ISSUE THREAD

**Issue by davidfowl** [aspnet/AspNetCore#5898]
As a follow up from https://github.com/aspnet/Hosting/issues/815, we want to allow specifying the path base as a host setting. Hosting should use this to call `UsePathBase` as the first middleware in the pipeline. IISIntegration does something similar with the `APPL_PATH ` env variable : https://github.com/aspnet/IISIntegration/blob/51554566532237ef76cb105ec5b3f7a5050b9ba4/src/Microsoft.AspNetCore.Server.IISIntegration/WebHostBuilderIISExtensions.cs#L19.

**Comment by shanselman** [comment]
ya this is annoyign we need to support the INVERSE

**Comment by shanselman** [comment]
I'm pretty sure this is the inverse of PathBase. I'm having to add it back in order to get ~ to work.

I'm taking http://staging.hanselman.com/blog and pointing it to http://hanselmanblog.azurewebsites.net so I need to add it BACK not strip it. Per https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.2#deal-with-path-base-and-proxies-that-change-the-request-path 

```
app.Use((context, next) =>
{
     context.Request.PathBase = new PathString(/blog);
     return next.Invoke();
});
```

**Comment by robertlestak** [comment]
I don't have a solution for the underlying AspNetCore issue, however I found a viable server-side workaround until a DotNet Core solution is found.

If a unique proxy_pass location block is created in the proxy server for every single DotNet Route, DotNet will be able to respond as expected.

This of course requires close coordination between server admins and app devs, or a script which dynamically modifies location blocks with each iterative update of the backing DotNet application.

While certainly is not a viable long-term solution, it does ensure that in the interim, while this bug is being addressed, DotNet APIs will "play nice" behind path-rewriting proxies.

**Comment by Andreas-Hjortland** [comment]
I encountered this issue today, and I solved it by adding a custom middleware which looks at the `X-Path-Base` header and sets the PathBase and Path based on the content in that (shamelessly copied the implementation from [UsePathBaseMiddleware.cs](https://github.com/dotnet/aspnetcore/blob/master/src/Http/Http.Abstractions/src/Extensions/UsePathBaseMiddleware.cs). Would it be possible to extend the [ForwardedHeadersMiddleware](https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs) and it corresponding Options class with something like this? I can create a pull request if this looks like a sound solution to the problem.

```c#
app.Use(async (context, next) =>
{
    // TODO Check if host is valid based on hosts in ForwardedHeadersOptions
    if (context.Request.Headers.TryGetValue("X-Path-Base", out var pathBase) &&
        context.Request.Path.StartsWithSegments(pathBase.Last(), out var matchedPath, out var remainingPath))
    {
        try
        {
            var originalPath = context.Request.Path;
            var originalPathBase = context.Request.PathBase;
            context.Request.PathBase = matchedPath;
            context.Request.Path = remainingPath;
            await next();
        }
        finally
        {
            context.Request.PathBase = new PathString(matchedPath);
            context.Request.Path = new PathString(remainingPath);
        }
    }
    else
    {
        await next();
    }
});
```

**Comment by ericis** [comment]
Just spent a few hours toward discovering this issue.

I'm using Kubernetes + Traefik reverse proxy, which strips the prefix path for two dockerized .NET sample apps. Of course, the static assets are attempting to use absolute paths at the root path, but when those URLs bounce through the proxy, they're not recognized.

e.g. "/app-1" on traefik strips the prefix to "/" for .NET and the sample app loves it, so it responds with a 200, but pukes out absolute URLs to the root "/", because it doesn't understand that it's running behind a proxy.

Naturally, I thought... this is a ubiquitous problem that has been solved using environment variables! wha, wha...

I "could" start coding, but why should I have to? I want to demo the awesome work of .NET Core teams out-of-the-box! 😉

Is there support yet for something like `ASPNETCORE_APPL_PATH` that I can pass to the docker container or for a header that I can inject using Traefik?

**Comment by ericis** [comment]
I'd love to see this:
`docker run --name aspnet-sample -d -p 8080:80 -e ASPNETCORE_APPL_PATH=/dotnet mcr.microsoft.com/dotnet/sample
s:aspnetapp`

... result in:

- 200: http://localhost:8080/dotnet *with static assets referenced at the URL "/dotnet"
- 404: http://localhost:8080/

**Comment by nsmithdev** [comment]
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1#deal-with-path-base-and-proxies-that-change-the-request-path

```
var appPath = Environment.GetEnvironmentVariable("ASPNETCORE_APPL_PATH");
if (!string.IsNullOrEmpty(appPath))
{
    app.Use((context, next) =>
    {
        context.Request.PathBase = new PathString(appPath);
        return next();
    });
}
```

**Comment by cslovell** [comment]
I'm genuinely surprised this isn't implemented. This is standard [12 factor app](https://12factor.net/config) stuff and for those of us in decoupled settings, where software built by one team is deployed internally in several locations, we need to be able to configure this from the environment without rebuilding the application. In other words, if the app is already pulling environment variables in, there should be a standard environment variable that we can use to set the base path. For instance: 

A .NET application built by one team is set to receive traffic at the root path, example.com. However, another team will deploy this application at example.com/team1/app, so they set up a proxy that forwards traffic and set the environment variable that instructs the application to preface all requests with this base path.

**Comment by cslovell** [comment]
> https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1#deal-with-path-base-and-proxies-that-change-the-request-path
> 
> ```
> var appPath = Environment.GetEnvironmentVariable("ASPNETCORE_APPL_PATH");
> if (!string.IsNullOrEmpty(appPath))
> {
>     app.Use((context, next) =>
>     {
>         context.Request.PathBase = new PathString(appPath);
>         return next();
>     });
> }
> ```

Yes - but this will need to be done for every application, despite being a standard component of 12 factor app development. The nature of organizational bureaucracy may imply that this change is not possible by the team deploying the software.

**Comment by adityamandaleeka** [comment]
Cleared the milestone so this shows up in our next triage discussion.

**Comment by adityamandaleeka** [comment]
Related: https://github.com/dotnet/aspnetcore/issues/23263 (doing that would give people another way to set the path base)

**Comment by adityamandaleeka** [comment]
Looks like this ancient issue could use some follow up... we added support for X-Forwarded-Prefix in the ForwardedHeadersMiddleware in .NET 8 (https://github.com/dotnet/aspnetcore/issues/23263), which allows specifying a path base for forwarded requests. If your application is running behind infrastructure that modifies the request path, you can configure the middleware to process this header using something like this:

```csharp
builder.Services.Configure<ForwardedHeadersOptions>(options =>
{
    options.ForwardedHeaders = ForwardedHeaders.XForwardedFor |
                               ForwardedHeaders.XForwardedProto |
                               ForwardedHeaders.XForwardedHost |
                               ForwardedHeaders.XForwardedPrefix;
});

app.UseForwardedHeaders();
```

With this, Request.PathBase will be correctly set based on X-Forwarded-Prefix. This resolves the issue for forwarded requests, which was the primary scenario discussed here.

If you've tried this out and it doesn't solve the issue for your scenario, please let us know.

**Comment by davidfowl** [comment]
Do all proxies set this? (it is the norm in nginx etc)

**Comment by shanselman** [comment]
Yes @davidfowl this is the standard and has been for years https://www.hanselman.com/blog/dealing-with-application-base-urls-and-razor-link-generation-while-hosting-aspnet-web-apps-behind-reverse-proxies and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For and https://www.loadbalancer.org/blog/nginx-and-x-forwarded-for-header/

**Comment by davidfowl** [comment]
I'm specifically talking about `X-Forwarded-Prefix`.

**Comment by shanselman** [comment]
@davidfowl my bad, yes X-Forwarded-Prefix is common for this reason https://josef.codes/using-the-x-forwarded-prefix-header-to-prefix-your-hateoas-links/ and is used to manage URL prefixes in reverse proxy setups, and is been around a while

**Comment by davidfowl** [comment]
I think closing the loop here might be updating our docs with how to configure this on popular reverse proxies (I'm not sure it's set by default in all the places). Doing a cursory glance I don't see top matches anywhere.

Also it seems like there are still cases where [explicit configuration is preferred](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-9.0#work-with-path-base-and-proxies-that-change-the-request-path)

**Comment by davidglassborow** [comment]
Just having the same issue, agree the docs might need an update.  In my case, we have a production site running in IIS using multiple app pools under a single unified url space, but want to be able to spin up the same code inside Azure Container Apps for staging, testing etc. 

Using nginix I used the following, and the asp ForwardedHeaders middleware then sorted out all the urls correctly

```
        location /admin/ {
            proxy_pass http://admin/;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            # nginx doesn't have a variable for the path, so need to hardcode
            proxy_set_header X-Forwarded-Prefix "/admin/";
        }
```
